src/on_shutdown.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY AUTHORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "on_shutdown.h"
+
+#include <stdlib.h>
+#include <small/rlist.h>
+#include <errno.h>
+
+struct on_shutdown_trigger {
+	struct trigger trigger;
+	/** Shutdown trigger function */
+	void (*handler)(void *);
+	/** Trigger function argument */
+	void *arg;
+};
+
+/**
+ * Shutdown trigger list, not need mutex, because access is
+ * avaliable only from tx thread
+ */
+static struct rlist on_shutdown_trigger_list =
+	RLIST_HEAD_INITIALIZER(on_shutdown_trigger_list);
+
+static int
+trigger_commom_f(struct trigger *trigger, MAYBE_UNUSED void *event)
+{
+	struct on_shutdown_trigger *on_shutdown_trigger =
+		container_of(trigger, struct on_shutdown_trigger, trigger);
+	on_shutdown_trigger->handler(on_shutdown_trigger->arg);
```

Comment:
I'd make handler return integer type so it can return 'return code'

Reply:
fix